### ISSUE TITLE
Explicit life cycle management

### ISSUE THREAD

**Issue by schuderer** [schuderer/mllaunchpad#17]
Right now, trained models are simply saved, there's one trained model per name-version combination (always the most recently trained one).

What a full life cycle of a model can look like:
1. Experiment/develop a model (outside of launchpad)
2. Implement first version of model in launchpad (by inheriting user model base classes)
3. Training in dev (version 0.1.0), checking results, trying it out (using clitool), maybe changing code once more (source code control is outside of launchpad's scope)
4. Creating release (same version number 0.1.0 for release itself and model, API inside), going through test/acceptance (or so). Uses separate config, model store, datasources/sinks per stage (dev/test/acc/prod or so). Training and retesting may be hooked up to model_actions instead of clitool now, API-app may be running on gunicorn/nginx or similar already (concrete env. is out of scope)
5. Publish to prod (again simply uses separate config with own prod model store and datasources/sinks). Training, testing, optional re-training should already be part of the environment by this stage. Regular or on-demand re-training overwrites the last model (a backup is created). (NOTE: testing evaluation currently would need to interpret the metrics directly. It might be worthwhile to make it possible for the user to make clear what metric should be used how to determine whether a model performs (a) well enough and (b) better/worse than another.
6. Working on a better model in DEV... Has a different version number (semantic, probably changed the middle number, e.g. 0.1.0 -> 0.2.0, because it's still compatible). Release, going through stages Test/Acc as before...
7. Found bug in PROD, branching from 0.1.0 release, fixing, incrementing version in respective branch from 0.1.0 -> 0.1.1, stages, release to PROD.
8. Back working on current DEV branch release, new release necessary as already created a release 0.2.0 earlier (now at 0.2.1), retesting, accepting...
9. Put model into production (again has its own config). This model 0.2.1 lives as a completely separate model next to version 0.1.0. This version 0.2.1 performs better than 0.1.0, but does not necessarily replace it.

Life cycle support looks actually quite complete up until point 9.

So either we have two instances running in parallel and some magic mechanism switches between them when requests come in (similar like load balancing but maybe based on how they perform). This seems wasteful of resources.

Or we always only load one of them (the better one) when launching the API-app. This means that the app needs to know how to decide which model to load. Currently it strictly loads only exactly the model-version combination given in the config.

So it needs to know (1) which candidate models are available in production (maybe simply the ones that are in the PROD model store) -- then we need to add model store functionality to unpublish (delete*) a specific model-version combination.

The API-app also needs to know (2) what versions of the same model are available in the model store and get their metrics -- need model store functionality to query for all model versions with a certain name (which can also be used to unpublish all versions in combination with the delete functionality).

The API-app also (3) needs to be able to decide what model actually is better (ways in config to do this).

These 3 points are actually general functionality that is just being used by the API-app.

What we also need (edit: "need" might be a strong word for some points):
- On testing, check not only whether a model is better/worse, but also whether it's better than a certain threshold. Issue a warning if not.
- A way to, after re-training, check whether the new model is actually really better than the previous one with the same 1.1.X version ("same" ignores the minor version) BEFORE publishing it to the model store (and giving a warning if it isn't because usually more recent data should be more relevant and lead to better performance).
- In general, we should, when identifying models by name+version for functional purposes, ignore the last, "fix", version number.
- Have a mechanism to look at published/trashed models with their current metrics visible, and republish/unpublish them
- A way to issue a warning if needed (scores are low, etc)
- If the major version of a new model changes, we do NOT simply replace the old model by default, but the default should be that it's treated as a completely different model (running in parallel).

*) "delete" means move to a backup location

**Comment by schuderer** [comment]
We probably also need to think about a management console in the browser or something similar.

**Comment by schuderer** [comment]
It appears that 1.0-level usage is also possible without any additional functionality of `mllaunchpad core`, so I'll move this out of the crucial items of the release-1.0 tracker for now. But it is still a relevant issue.
